core/src/main/scala/org/apache/spark/TaskContextImpl.scala

```diff
@@ -140,16 +140,22 @@ private[spark] class TaskContextImpl(
   }
 
   /** Marks the task for interruption, i.e. cancellation. */
-  private[spark] def markInterrupted(): Unit = {
-    interrupted = true
+  private[spark] def markInterrupted(reason: String): Unit = {
+    maybeKillReason = Some(reason)
+  }
+
+  private[spark] override def killTaskIfInterrupted(): Unit = {
+    if (maybeKillReason.isDefined) {
+      throw new TaskKilledException(maybeKillReason.get)
```

Comment:
This is not thread safe - while technically we do not allow kill reason to be reset to None right now and might be fine, it can lead to future issues.

Either make all access/updates to kill reason synchronized; or capture `maybeKillReason` to a local variable and use that in the `if` and `throw`